include/dali/core/event_pool.h

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_EVENT_POOL_H_
+#define DALI_CORE_EVENT_POOL_H_
+
+#include <vector>
+#include <utility>
+#include "dali/core/cuda_event.h"
+#include "dali/core/spinlock.h"
+
+namespace dali {
+
+class DLL_PUBLIC CUDAEventPool {
+ public:
+  ~CUDAEventPool();
+  explicit CUDAEventPool(unsigned event_flags = cudaEventDisableTiming);
+
+  /**
+   * @brief Get an event for given device.
+   *
+   * @param device_id   CUDA runtime API device ordinal. If negative, calling thread's
+   *                    current device is used.
+   *
+   * @return A CUDA event wrapper object. If there were any events in the pools, the
+   *         event is taken from it, otherwise a new event is created.
+   */
+  CUDAEvent Get(int device_id = -1);
+
+  /**
+   * @brief Place an event for given device in the pool.
+   *
+   * @param event     CUDA event wrapper object. The caller relinquishes ownership of the event.
+   * @param device_d  CUDA runtime API device ordinal of the device for which the event was
+   *                  created. If negative, calling thread's current device is used.
+   *
+   * @remarks It is an error to misstate the device_id. Placing an event with improper deviceid
```

Comment:
Maybe CUDAEvent should keep the device_id as a field?

Reply:
I thought about it. It would lose the semantics of `UniqueHandle<cudaEvent_t>` (and it would grow twice in size), but maybe it's worth it.